internal/reminder/reminder.go

```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
 		return errors.New("reminder stopped, cannot start again")
 	default:
 	}
+	defer r.Stop()
 
 	interval := r.cfg.RecurrenceConfig.Interval
 	if interval <= 0 {
 		return fmt.Errorf("invalid interval: %s", r.cfg.RecurrenceConfig.Interval)
 	}
 
+	metricsServerDone := make(chan struct{})
+
+	if r.cfg.MetricsConfig.Enabled {
+		metricsProviderReady := make(chan struct{})
+
+		go func() {
+			if err := r.startMetricServer(ctx, metricsProviderReady); err != nil {
+				logger.Fatal().Err(err).Msg("failed to start metrics server")
+			}
+			close(metricsServerDone)
+		}()
+
+		select {
+		case <-metricsProviderReady:
+			var err error
+			r.metrics, err = metrics.NewMetrics(otel.Meter("reminder"))
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			logger.Info().Msg("reminder stopped")
+			return nil
+		}
+	}
+
 	r.ticker = time.NewTicker(interval)
-	defer r.Stop()
 
 	for {
 		select {
 		case <-ctx.Done():
+			if r.cfg.MetricsConfig.Enabled {
+				<-metricsServerDone
+			}
```

Comment:
Same for this one (making sure I'm not complicating things)

Reply:
Here, you might more easily do the following:

```golang
if r.cfg.MetricsConfig.Enabled {
  ...
} else {
  close(metricsServerDone)
}
```

And then you won't need the conditional here.

Reply:
I didn't understand that. The current shutdown sequence waits for the metrics server to be done. (waiting so that deferred call cannot be executed i.e. the `Stop()` call for reminder)

Reply:
If you `close(metricsServerDone)` in the `else` block on 111, 118-120 and 124-126 can simply do `<-metricsServerDone` whether or not `r.cfg.MetricsConfig.Enabled` is true or false, because either case will close the channel.

I'd also be okay with leaking the metrics HTTP server if it doesn't cause harm elsewhere.

Reply:
Yeah, good point. I implemented this change.